File: shared/auth/rbac/scope.ts
// Scope predicate functions (M02 PLAN §3.3).
// These are pure functions — zero I/O.

import type { AuthContext, ScopeContext } from './can.js';

/**
 * group scope: campaignId must be in authCtx.allowedCampaigns.
 * null/undefined campaignId → deny (fail-closed).
 * allowedCampaigns === '*' → always pass.
 */
export function passGroupScope(auth: AuthContext, scope: ScopeContext): boolean {
  if (scope.campaignId === undefined) return false;
  if (auth.allowedCampaigns === '*') return true;
  // allowedCampaigns === null (no group) treated as empty array → deny
  if (!Array.isArray(auth.allowedCampaigns)) return false;
  return auth.allowedCampaigns.includes(scope.campaignId);
}

/**
 * own scope: ownerUserId === actor uid, OR uid in assignedTo list.
 * Missing ownerUserId and missing assignedTo → deny.
 */
export function passOwnScope(auth: AuthContext, scope: ScopeContext): boolean {
  if (scope.ownerUserId !== undefined && scope.ownerUserId === auth.uid) return true;
  if (Array.isArray(scope.assignedTo) && scope.assignedTo.includes(auth.uid)) return true;
  return false;
}

/**
 * self scope: targetUserId must equal the actor's uid.
 * Missing targetUserId → deny.
 */
export function passSelfScope(auth: AuthContext, scope: ScopeContext): boolean {
  return scope.targetUserId !== undefined && scope.targetUserId === auth.uid;
}

File: dialer/internal/auth/rbac/scope.go
// Scope predicate functions — Go mirror of shared/auth/rbac/scope.ts (M02 PLAN §3.3).
package rbac

// PassGroupScope returns true when campaignID is in auth.AllowedCampaigns
// (or AllCampaigns == true). Zero campaignID is treated as not-set → deny.
func PassGroupScope(auth AuthContext, scope ScopeContext) bool {
	if scope.CampaignID == 0 {
		return false
	}
	if auth.AllCampaigns {
		return true
	}
	for _, id := range auth.AllowedCampaigns {
		if id == scope.CampaignID {
			return true
		}
	}
	return false
}

// PassOwnScope returns true when ownerUserID matches auth.UID
// OR auth.UID appears in the assignedTo list.
func PassOwnScope(auth AuthContext, scope ScopeContext) bool {
	if scope.OwnerUserID != 0 && scope.OwnerUserID == auth.UID {
		return true
	}
	for _, id := range scope.AssignedTo {
		if id == auth.UID {
			return true
		}
	}
	return false
}

// PassSelfScope returns true when targetUserID equals auth.UID.
// Zero targetUserID → deny.
func PassSelfScope(auth AuthContext, scope ScopeContext) bool {
	return scope.TargetUserID != 0 && scope.TargetUserID == auth.UID
}
